File: indico/modules/events/registration/placeholders/invitations.py
# This file is part of Indico.
# Copyright (C) 2002 - 2024 CERN
#
# Indico is free software; you can redistribute it and/or
# modify it under the terms of the MIT License; see the
# LICENSE file for more details.

from markupsafe import Markup

from indico.util.i18n import _
from indico.util.placeholders import Placeholder
from indico.web.flask.util import url_for


class FirstNamePlaceholder(Placeholder):
    name = 'first_name'
    description = _('First name of the person')

    @classmethod
    def render(cls, invitation, **kwargs):
        return invitation.first_name


class LastNamePlaceholder(Placeholder):
    name = 'last_name'
    description = _('Last name of the person')

    @classmethod
    def render(cls, invitation, **kwargs):
        return invitation.last_name


class EmailPlaceholder(Placeholder):
    name = 'email'
    description = _('Email of the person')

    @classmethod
    def render(cls, invitation, **kwargs):
        return invitation.email


class InvitationLinkPlaceholder(Placeholder):
    name = 'invitation_link'
    description = _('Link to accept/decline the invitation')
    required = True

    @classmethod
    def render(cls, invitation, **kwargs):
        url = url_for('.display_regform', invitation.locator.uuid, _external=True)
        return Markup(f'<a href="{url}">{url}</a>')

File: indico/modules/events/registration/client/js/invitations.js
// This file is part of Indico.
// Copyright (C) 2002 - 2025 CERN
//
// Indico is free software; you can redistribute it and/or
// modify it under the terms of the MIT License; see the
// LICENSE file for more details.

/* eslint-disable import/unambiguous */
/* global handleAjaxError:false */

import React from 'react';
import ReactDOM from 'react-dom';

import {Translate} from 'indico/react/i18n';
import {camelizeKeys} from 'indico/utils/case';

import EmailPendingInvitationsButton from './components/EmailPendingInvitationsButton';
import InvitationList from './components/InvitationList';
import InviteDialogButton from './components/InviteDialogButton';

(function(global) {
  global.setupInvitationPage = function setupInvitationPage({
    eventId,
    regformId,
    hasPendingInvitations,
    invitations,
  }) {
    $('#invitation-list-container').on('indico:confirmed', '.js-invitation-action', function(evt) {
      evt.preventDefault();

      const $this = $(this);
      $.ajax({
        url: $this.data('href'),
        method: $this.data('method'),
        complete: IndicoUI.Dialogs.Util.progress(),
        error: handleAjaxError,
        success(data) {
          renderInvitationPage({
            eventId,
            regformId,
            hasPendingInvitations: data.has_pending_invitations,
            invitations: data.invitation_list,
          });
        },
      });
    });

    const handleInvitationsChanged = data => {
      renderInvitationPage({
        eventId,
        regformId,
        hasPendingInvitations: data.has_pending_invitations,
        invitations: data.invitation_list,
      });
    };

    renderInviteButton({eventId, regformId, onInvitationsChanged: handleInvitationsChanged});
    renderInvitationPage({eventId, regformId, hasPendingInvitations, invitations});
  };

  function renderEmailInvitationsBtn({eventId, regformId, hasPendingInvitations}) {
    const container = document.getElementById('email-all-pending-invitations-container');
    ReactDOM.render(
      <EmailPendingInvitationsButton
        eventId={eventId}
        regformId={regformId}
        label={Translate.string('Remind all')}
        disabled={!hasPendingInvitations}
      />,
      container
    );
  }

  function renderInvitationsList({eventId, regformId, invitations}) {
    const container = document.getElementById('invitation-list-container');
    ReactDOM.render(
      <InvitationList
        eventId={eventId}
        regformId={regformId}
        invitations={camelizeKeys(invitations)}
      />,
      container
    );
  }

  function renderInviteButton({eventId, regformId, onInvitationsChanged}) {
    const container = document.getElementById('invite-dialog-button-container');
    if (!container) {
      return;
    }
    ReactDOM.render(
      <InviteDialogButton
        eventId={eventId}
        regformId={regformId}
        onInvitationsChanged={onInvitationsChanged}
      />,
      container
    );
  }

  function renderInvitationPage({eventId, regformId, hasPendingInvitations, invitations}) {
    renderEmailInvitationsBtn({
      eventId,
      regformId,
      hasPendingInvitations,
    });
    renderInvitationsList({
      eventId,
      regformId,
      invitations,
    });
  }
})(window);
